### PR TITLE
Improved error handling when using multiple tilemap collision objects in a single game object

### LINKED ISSUE #6386: tilemap.set_tile(map_id, layer, tile_x, tile_y, 0) can crash under certain circumstances
**Describe the bug (REQUIRED)**
Under certain circumstances, erasing tiles in a tilemap using tilemap.set_tile(map_id, layer, tile_x, tile_y, 0) can cause the game to crash.

**To Reproduce (REQUIRED)**
*Precise* repro steps are unknown, but I've narrowed it down to a pretty small repro case project, included.

**Expected behavior (REQUIRED)**
It should not crash. Worst case, it should log an error, but the error should point to what game code caused the error – as it is, the game doesn't log anything useful.

**Defold version (REQUIRED):**
  - Version 1.2.192

**Platforms (REQUIRED):**
 - Platforms: macOS
 - OS: macOS 11.6.2

**Minimal repro case project (OPTIONAL):**
Please attach a minimal project where the bug is reproduced. This will greatly help the person trying to investigate and fix the bug.
[tiled-test 2.zip](https://github.com/defold/defold/files/8045209/tiled-test.2.zip)

**Logs (OPTIONAL):**
```
ERROR:GAMESYS: SetGridShapeHull: <row,column> out of bounds
Assertion failed: (fixture != 0x0), function GetFixture, file physics_2d.cpp, line 655.
INFO:CRASH: Successfully wrote Crashdump to file: /Users/greay/Library/Application Support/Defold/_crash
ERROR:CRASH: CALL STACK:

# 0 pc      0x91d7d libc++abi.dylib _sigtramp+29
```

**Workaround (OPTIONAL):**
n/a

**Additional context (OPTIONAL):**
While all of the extra logic has been stripped out, the purpose of the code is to replace certain tiles on the map with animated characters or objects.

### PR DESCRIPTION
**Pull request by britzl**
When using `tilemap.set_tile()` we currently broadcast a message to all collision components on the same game object that a tile was changed. This is potentially problematic if there are multiple collision components referencing different tilemaps as their shapes because all collision components will react to the message and update, regardless if they used the same tilemap shape or not. What's worse is if the tilemaps have different number of layers it will in fact result in a crash.

This kind of very loose coupling between tilemaps and collision components is unfortunate and does only really work well in a scenario where you have ONE tilemap and ONE collision component using the tilemap as its shape. Solving this design flaw is not covered by this fix which only adds improved error handling to avoid a crash.

Fixes #6386

### CODE REVIEW

File: engine/physics/src/physics/physics_2d.cpp [hunk 1]
```diff
@@ -654,39 +654,42 @@ namespace dmPhysics
...
+    bool SetGridShapeHull(HCollisionObject2D collision_object, uint32_t shape_index, uint32_t row, uint32_t column, uint32_t hull, HullFlags flags)
     {
         b2Body* body = (b2Body*) collision_object;
-        b2GridShape* grid_shape = GetGridShape(body, shape_index);
+        b2Fixture* fixture = GetFixture(body, shape_index);
+        if (fixture == 0)
+        {
+            return false;
+        }
+        b2GridShape* grid_shape = (b2GridShape*) fixture->GetShape();
```
Comment (JCash) [inline engine/physics/src/physics/physics_2d.cpp:668]: You removed the "GetGridShape()" which actually checks the type.
How do you know this is a grid shape?

File: engine/physics/src/physics/test/test_physics_2d.cpp [hunk 2]
```diff
@@ -203,6 +203,41 @@ TYPED_TEST(PhysicsTest, UseBullet)
...
+                                    0.5f,  0.5f,
+                                   -0.5f,  0.5f};
+
+    const dmPhysics::HullDesc hulls[] = { {0, 4}, {0, 4} };
+    dmPhysics::HHullSet2D hull_set = dmPhysics::NewHullSet2D(TestFixture::m_Context, hull_vertices, 4, hulls, 2);
+    dmPhysics::HCollisionShape2D grid_shape = dmPhysics::NewGridShape2D(TestFixture::m_Context, hull_set, Point3(0,0,0), cell_width, cell_height, rows, columns);
+    typename TypeParam::CollisionObjectType grid_co = (*TestFixture::m_Test.m_NewCollisionObjectFunc)(TestFixture::m_World, data, &grid_shape, 1u);
+
+    ASSERT_TRUE(dmPhysics::SetGridShapeHull(grid_co, 0, 0, 0, 0, EMPTY_FLAGS));
+    ASSERT_FALSE(dmPhysics::SetGridShapeHull(grid_co, 1, 0, 0, 0, EMPTY_FLAGS));
```
Comment (JCash) [inline engine/physics/src/physics/test/test_physics_2d.cpp:233]: Coming back to the question above, if the shape at index 1 was a box shape, what would happen?

**Review by JCash — CHANGES_REQUESTED** [review]

File: engine/physics/src/physics/physics_2d.cpp (hunk 1 above)
Comment (britzl) [reply engine/physics/src/physics/physics_2d.cpp:668]: Oh crap, didn't notice this. Will update!

File: engine/physics/src/physics/physics_2d.cpp [hunk 3]
```diff
@@ -654,39 +654,42 @@ namespace dmPhysics
...
-    void SetGridShapeEnable(HCollisionObject2D collision_object, uint32_t shape_index, uint32_t enable)
+    bool SetGridShapeEnable(HCollisionObject2D collision_object, uint32_t shape_index, uint32_t enable)
     {
         b2Body* body = (b2Body*) collision_object;
         b2Fixture* fixture = GetFixture(body, shape_index);
+        if (fixture == 0)
+        {
+            return false;
+        }
         b2GridShape* grid_shape = (b2GridShape*) fixture->GetShape();
```
Comment (britzl) [inline engine/physics/src/physics/physics_2d.cpp:685]: Why are we not checking the same thing here though with regard to if the shape is actually a b2GridShape or not?

Comment (JCash) [reply engine/physics/src/physics/physics_2d.cpp:694]: Just oversight I'd guess.

Comment (JCash) [reply engine/physics/src/physics/physics_2d.cpp:694]: To be clear, it is also a bug

Comment (britzl) [reply engine/physics/src/physics/physics_2d.cpp]: Updated PR with fix for this

**Review by JCash — APPROVED** [review]

**Pull request merged**